README.md
# Router

![PHP Support](https://img.shields.io/badge/php%20tested-5.6-brightgreen.svg)
![PHP Support](https://img.shields.io/badge/php%20tested-7.1-brightgreen.svg)
![Dependencies](https://img.shields.io/badge/dependencies-none-brightgreen.svg)
![License](https://img.shields.io/badge/license-MIT-green.svg)
![Travis](https://api.travis-ci.org/GonistLelatel/aphp_router.svg?branch=master)

## Introduction

`aphp/router` is a lightweight and simple object oriented PHP Router.

## Installation
PHP5.6 , PHP7.0+

`composer require aphp/router`

## Features

* Supports GET, POST, PUT, DELETE, OPTIONS, PATCH and HEAD request methods.
* Routing shorthands such as get(), post(), put(), …
* Dynamic Route Patterns: PCRE-based.
* Route groups.
* Supports X-HTTP-Method-Override header.
* 404 handling.

## Syntax
Intialization
```php
$router = new Router();

$router->set404(function() use ($router) {
	header($_SERVER["SERVER_PROTOCOL"] . " 404 Not Found");
	// ...
});

// ...

$router->run();
```
Shorthands for single request methods are provided:
```php
$router->get('pattern', function() { /* ... */ });
$router->post('pattern', function() { /* ... */ });
$router->put('pattern', function() { /* ... */ });
$router->delete('pattern', function() { /* ... */ });
$router->options('pattern', function() { /* ... */ });
$router->patch('pattern', function() { /* ... */ });
```
PCRE-based Route Patterns

* `/movies/(\d+)` = One or more digits (0-9)
* `/profile/(\w+)` = One or more word characters (a-z 0-9 _)
* `[a-z0-9_-]+`  = One or more word characters (a-z 0-9 _) and the dash (-)
* `.*?` = Any character (including /), zero or more
* `[^/]+` = Any character but /, one or more
* `{param}` = Equivalent `(.*?)`

Note: [PHP-regex-cheat-sheet.pdf](https://courses.cs.washington.edu/courses/cse190m/12sp/cheat-sheets/php-regex-cheat-sheet.pdf)
### Groups
Use $router->group($baseroute, $fn) to group a collection of routes onto a subroute pattern.
```php
$router->group('/movies', function() use ($router) {
    // will result in '/movies/'
    $router->get('/', function() {
        echo 'movies overview';
    });
    // will result in '/movies/id'
    $router->get('/(\d+)', function($id) {
        echo 'movie id ' . htmlentities($id);
    });
});
```
### Cancel routing

In some cases needs to cancel routing, by the application logic.<br>
Use `$router->cancel()` method to handle this.

```php 
$router->set404(function() use ($router) {
	header($_SERVER["SERVER_PROTOCOL"] . " 404 Not Found");
	echo '404 Not Found';
});
$router->get('/movies/(\d+)', function($id) use($router) {
	if ($id == 5000) {
		$router->cancel(); // <<
		return;
	}
	echo 'movie id ' . htmlentities($id);
});
$router->run();
```
If `\movies\5000` then 404 page will show.

## More features
For more features:
* Read source code and examples
* Practice with `Router` in real code


CHANGELOG.md
# Router changelog

## Future
* Documentation, test and code cleanup.

## 1.0.1

* Added: cancel method.

## 1.0.0
* First release.

tests/startTest.sh
#!/bin/bash
#
# install phpunit before tests
# https://phpunit.de/getting-started/phpunit-5.html
#
# cd /usr/local/bin
# dir
# sudo wget -O phpunit https://phar.phpunit.de/phpunit-5.phar
# sudo chmod +x phpunit
#
# run webserver before test
# startServer.sh

BASEDIR=$(dirname "$0")
cd $BASEDIR
phpunit --bootstrap bootstrap_autoload.php . --debug

tests/RouterTest.php
<?php 

class RouterTest extends Base_TestCase {
	
	// STATIC
	public static function setUpBeforeClass() {
		
	}

	public static function tearDownAfterClass() {
		
	}
	
	// override
	
	protected function setUp() {
		
	}
	
	protected function tearDown() {

	}
	
	protected function loadServerRequest($rq) {
		$c = file_get_contents('http://localhost:8008' . $rq);
		return $c;
	}
	
	// tests
	
	public function test_1() 
	{
		$t = $this->loadServerRequest('/movies/name/photos/12');
		$this->assertEquals($t , 'get /movies/(\w+)/photos/(\d+) = name, 12');

		$t = $this->loadServerRequest('/movies');
		$this->assertEquals($t , 'movies overview');

		$t = $this->loadServerRequest('/movies/');
		$this->assertEquals($t , 'movies overview');

		$t = $this->loadServerRequest('/movies/all/man');
		$this->assertEquals($t , 'group 2 : /movies/all/man');

		$t = $this->loadServerRequest('/movies/all/man/12');
		$this->assertEquals($t , 'group 2 : /movies/all/(\d+) = 12');

		$t = $this->loadServerRequest('/movies/car/woman');
		$this->assertEquals($t , 'group 1 : /movies/car/woman');

		$t = $this->loadServerRequest('/photos/all');
		$this->assertEquals($t , 'group photos : /photos/all');

		$t = $this->loadServerRequest('/simplelastHelloWorld');
		$this->assertEquals($t , 'get simplelastHelloWorld');
	}
}
	
	

tests/startServer.sh
#!/bin/bash
#
# run webserver before test
# startServer.sh

BASEDIR=$(dirname "$0")
cd $BASEDIR
php -S localhost:8008 -t webserver webserver/routing.php

src/Request.php
<?php 

namespace aphp\Router;

class Request {
	public $headers = null;
	public $method = null;
	public $uri = null;
	/**
     * Get all request headers.
     *
     * @return array The request headers
     */
	public function getRequestHeaders() {
		if ($this->headers) {
			return $this->headers;
		}
		$headers = [];
        // If getallheaders() is available, use that
        if (function_exists('getallheaders')) {
            $headers = getallheaders();

            // getallheaders() can return false if something went wrong
            if ($headers !== false) {
				$this->headers = $headers;
                return $headers;
            }
        }
        // Method getallheaders() not available or went wrong: manually extract 'm
        foreach ($_SERVER as $name => $value) {
            if ((substr($name, 0, 5) == 'HTTP_') || ($name == 'CONTENT_TYPE') || ($name == 'CONTENT_LENGTH')) {
                $headers[str_replace([' ', 'Http'], ['-', 'HTTP'], ucwords(strtolower(str_replace('_', ' ', substr($name, 5)))))] = $value;
            }
		}
		$this->headers = $headers;
        return $headers;
	}

	/**
     * Get the request method used, taking overrides into account.
     *
     * @return string The Request method to handle
     */
	public function getRequestMethod()
    {
		if ($this->method) {
			return $this->method;
		}
		// Take the method as found in $_SERVER
        $method = $_SERVER['REQUEST_METHOD'];

        // If it's a HEAD request override it to being GET and prevent any output, as per HTTP Specification
        // @url http://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9.4
        if ($_SERVER['REQUEST_METHOD'] == 'HEAD') {
			//ob_start(); // Router cant stop echo operations
			// Web application is the main logic, not router
            $method = 'GET';
        }

        // If it's a POST request, check for a method override header
        elseif ($_SERVER['REQUEST_METHOD'] == 'POST') {
            $headers = $this->getRequestHeaders();
            if (isset($headers['X-HTTP-Method-Override']) && in_array($headers['X-HTTP-Method-Override'], ['PUT', 'DELETE', 'PATCH'])) {
                $method = $headers['X-HTTP-Method-Override'];
            }
        }
		$this->method = $method;
        return $method;
	}
	/**
     * Define the current relative URI.
     *
     * @return string
     */
	public function getCurrentUri()
    {
		if ($this->uri) {
			return $this->uri;
		}
		// Get the current Request URI and remove rewrite base path from it (= allows one to run the router in a sub folder)
        $uri = substr(rawurldecode($_SERVER['REQUEST_URI']), strlen($this->getBasePath()));

        // Don't take query params into account on the URL
        if (strstr($uri, '?')) {
            $uri = substr($uri, 0, strpos($uri, '?'));
        }

		// Remove trailing slash + enforce a slash at the start
		$this->uri = '/' . trim($uri, '/');
        return $this->uri;
	}
	
	public function getBasePath()
    {
		// override if needed
		// return implode('/', array_slice(explode('/', $_SERVER['SCRIPT_NAME']), 0, -1)) . '/';
		return '';
	}
}

src/BaseRouter.php
<?php 

namespace aphp\Router;

/*
$router->get('/movies/(\d+)/photos/(\d+)', function($movieId, $photoId) {
    echo 'Movie #' . $movieId . ', photo #' . $photoId);
});
*/

// -----------
// RouterH
// -----------

abstract class BaseRouterH {
	/**
     * if true , then $pattern = RAW preg pattern
     */
	public $useRawPattern = false;
	public $request = null; // Request
	public $currentPattern = null;
	public $finished = false;
	
	/**
     * Set the 404 handling function.
     */
	abstract public function set404($fn);

	/**
     * Groups a collection of callbacks onto a base route.
     *
     * @param string $baseRoute The route sub pattern to mount the callbacks on
     * @param callable $fn        The callback method
     */
	abstract public function group($pattern, $fn);

	 /**
     * Store a route and a handling function to be executed when accessed using one of the specified methods.
     *
     * @param array  $methods [string] methods
     * @param string $pattern A route pattern such as /movies/(\d+)/photos/(\d+)
     * @param object|callable $fn The handling function to be executed
     */
	abstract public function match($methods, $pattern, $fn);

	/**
	 * Runs 404 if needed 
	 * Last method of the router
	*/
	abstract public function run();
}

// -----------
// Router
// -----------

class BaseRouter extends BaseRouterH {
	public function __construct(Request $request = null) {
		if (!$request) {
			$this->request = new Request();
		} else {
			$this->request = $request;
		}
	}

	public function set404($fn) {
		$this->fn_404 = $fn;
	}

	public function group($pattern, $fn) {
		$this->group[] = $this->patternString($pattern);
		if (is_callable($fn)) {
			call_user_func($fn);
		}
		array_pop($this->group);
	}

	public function match($methods, $pattern, $fn) {
		if ($this->finished) {
			return;
		}
		if (in_array($this->request->getRequestMethod(), $methods) === false) {
			return;
		}
		$pattern = implode('', $this->group) . $this->patternString($pattern);
		if ($pattern !== '/') {
			$pattern = rtrim($pattern, '/');
		}
		$this->currentPattern = $pattern;

		// Replace all curly braces matches {} into word patterns (like Laravel)
		$pattern = preg_replace('~{(.*?)}~', '(.*?)', $pattern);

		if (!$this->useRawPattern) {
			$pattern = '~^' . $pattern . '$~';
		}

		if (preg_match($pattern, $this->request->getCurrentUri(), $matches)) {
			$this->finished = true;
			$params = array_slice($matches, 1);
			if (is_callable($fn)) {
				call_user_func_array($fn, $params);
			}
		}
	}

	public function run() {
		if ($this->finished) {
			return;
		}
		if (is_callable($this->fn_404)) {
			call_user_func($this->fn_404);
		}
	}

	// Protected

	protected $fn_404;
	protected $group = [ ];

	protected function patternString($pattern) {
		return '/' . trim($pattern, '/');
	}
}

example/example00.php
<?php

require __DIR__ . '/../vendor/autoload.php';

use aphp\Router\Router;
use aphp\Router\Request;

// /movies/(\w+)/photos/(\w+)
// /movies/(\d+)/photos/(\d+)
// /movies/(\d+)/photos/(\d+)

$request = new Request();
//$request->uri = '/movies/ssdsd/photos/100'; // edit this line
//$request->uri = '/movies/5000'; // edit this line
$request->uri = '/movies/5001'; // edit this line
$request->method = 'GET';

$router = new Router($request);

$router->set404(function() use ($router) {
	//header($_SERVER["SERVER_PROTOCOL"] . " 404 Not Found");
	echo '404 Not Found';
});

$router->match(['GET'], '/movies/(\w+)/photos/(\d+)', function($p1, $p2) use ($router) {
	echo '/movies/(\w+)/photos/(\d+) =  ' . $p1 . ' ' . $p2;
});

$router->get('/movies/(\d+)', function($id) use($router) {
	if ($id == 5000) {
		echo 'cancel detected' . PHP_EOL;
		$router->cancel(); // <<
		return;
	}
	echo 'movie id ' . htmlentities($id);
});

$router->run();



